### PR TITLE
Add conformance tests for free()

### PR DESCRIPTION
**Pull request by padmashree9110**
Adds
1> 1-1 test
2> assertions.xml
3> coverage.txt

Change-Id: I1b8c203c259a96bec5446aa305e78c00170758a4
Signed-off-by: Mandri, Padmashree <padmashree.mandri@intel.com>

### CODE REVIEW

File: testcases/open_posix_testsuite/conformance/interfaces/free/1-1.c [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+#include "posixtest.h"
+
+#define TNAME "free/1-1.c"
+
+int main(void)
+{
+	int i = 0;
+	struct rlimit rlim;
+	char *ptr1 = NULL;
+	char *ptr2 = NULL;
```
Comment (metan-ucw) [inline testcases/open_posix_testsuite/conformance/interfaces/free/1-1.c]: What is the point of initializing variables that are never used before they are assigned.

Also for return values the most common name is ret not i.

File: testcases/open_posix_testsuite/conformance/interfaces/free/1-1.c [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+			free(ptr2);
+			exit(PTS_PASS);
+		}
+	} else {
+		printf(TNAME " Test Failed, malloc() is able to allocate memory before free(),"
+				" errno: %d.\n", errno);
+		free(ptr2);
+		free(ptr1);
+		exit(PTS_FAIL);
+	}
```
Comment (metan-ucw) [inline testcases/open_posix_testsuite/conformance/interfaces/free/1-1.c]: Why can't we do it as:

```
if (ptr2 == NULL) {
     ...
    exit(PTS_FAIL);
}

free(ptr1);
ptr2=malloc(...)
...
```

That's way easier to read than the if-then-else spaghetti you have here.

File: testcases/open_posix_testsuite/conformance/interfaces/free/coverage.txt [hunk 3]
```diff
@@ -0,0 +1,8 @@
+This file defines the coverage for the free() function testing.
+
+	Assertion	Status
+	1			YES
+	2			NO *
+
+	* - Hard to test
```
Comment (metan-ucw) [inline testcases/open_posix_testsuite/conformance/interfaces/free/coverage.txt]: Well you can at least do a test that calls free(NULL); exit(0); which does not cover much, but is still better than nothing.

Comment (padmashree9110) [reply testcases/open_posix_testsuite/conformance/interfaces/free/coverage.txt]: okie, added test in new patch.

File: testcases/open_posix_testsuite/conformance/interfaces/free/1-1.c (hunk 2 above)
Comment (padmashree9110) [reply testcases/open_posix_testsuite/conformance/interfaces/free/1-1.c]: Done

File: testcases/open_posix_testsuite/conformance/interfaces/free/1-1.c (hunk 1 above)
Comment (padmashree9110) [reply testcases/open_posix_testsuite/conformance/interfaces/free/1-1.c]: Updated in new patch.

**Comment by padmashree9110** [comment]
@metan-ucw Please review/merge the patch.

File: testcases/open_posix_testsuite/conformance/interfaces/free/1-1.c [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+ *	-Allocate memory to pointer_1 of size 1024 * 1024 *1.
+ *	-Alloacte memory for pointer_2 of size 1024 *1024 *1.
+ *	But it will fail due to insufficient memory.
+ *	-Try allocating memory to pointer_2 after freeing memory
+ *	allocated to pointer_1.
+ *	-Check whether memory is allocated by malloc().
+ *
+*/
+
+#define _XOPEN_SOURCE 700
```
Comment (metan-ucw) [inline testcases/open_posix_testsuite/conformance/interfaces/free/1-1.c]: The explicit definitions of XOPEN_SOURCE were removed in 310c9eb6e6af321c7515b1fc59df73c2a5b7c084 new tests must not include these.

File: testcases/open_posix_testsuite/conformance/interfaces/free/1-1.c [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+#include <stdlib.h>
+#include <string.h>
+#include <sys/resource.h>
+#include "posixtest.h"
+
+#define TNAME "free/1-1.c"
+
+int main(void)
+{
+	int ret = 0;
```
Comment (metan-ucw) [inline testcases/open_posix_testsuite/conformance/interfaces/free/1-1.c]: This is useless initialization.

File: testcases/open_posix_testsuite/conformance/interfaces/free/1-1.c [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+		printf(TNAME " Error at setrlimit(), errno: %d\n", errno);
+		exit(PTS_UNRESOLVED);
+	}
+
+	errno = 0;
+	ptr1 = malloc(1024 * 1024 * 1);
+	/* Malloc needs some memeory for its own data structure so more than
+	1024 * 1024 * 1 is used here */
+	if (ptr1 == NULL) {
+		printf(TNAME " Malloc failed(), errno: %d\n", errno);
```
Comment (metan-ucw) [inline testcases/open_posix_testsuite/conformance/interfaces/free/1-1.c]: Should be "malloc() failed" there is no failed() function in POSIX.

File: testcases/open_posix_testsuite/conformance/interfaces/free/1-1.c [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+	errno = 0;
+	ret = setrlimit(RLIMIT_DATA, &rlim);
+	if (ret == -1) {
+		printf(TNAME " Error at setrlimit(), errno: %d\n", errno);
+		exit(PTS_UNRESOLVED);
+	}
+
+	errno = 0;
+	ptr1 = malloc(1024 * 1024 * 1);
```
Comment (metan-ucw) [inline testcases/open_posix_testsuite/conformance/interfaces/free/1-1.c]: The * 1 is useless.

File: testcases/open_posix_testsuite/conformance/interfaces/free/1-1.c [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+	/* Malloc needs some memeory for its own data structure so more than
+	1024 * 1024 * 1 is used here */
+	if (ptr1 == NULL) {
+		printf(TNAME " Malloc failed(), errno: %d\n", errno);
+		exit(PTS_UNRESOLVED);
+	}
+
+	errno = 0;
+	ptr2 = malloc(1024 * 1024 * 1);
```
Comment (metan-ucw) [inline testcases/open_posix_testsuite/conformance/interfaces/free/1-1.c]: Here as well.

File: testcases/open_posix_testsuite/conformance/interfaces/free/1-1.c [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+	if (ptr2 != NULL) {
+		printf(TNAME " Test Failed, malloc() is able to allocate memory before free(),"
+				" errno: %d.\n", errno);
```
Comment (metan-ucw) [inline testcases/open_posix_testsuite/conformance/interfaces/free/1-1.c]: This message is way too long. Could be "Test Failed: malloc() allocated memory before free()" as there is no point to print errno on success.

File: testcases/open_posix_testsuite/conformance/interfaces/free/1-1.c [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+		exit(PTS_FAIL);
+	}
+	free(ptr1);
+	errno = 0;
+	ptr2 = malloc(1024 * 1024 * 1);
```
Comment (metan-ucw) [inline testcases/open_posix_testsuite/conformance/interfaces/free/1-1.c]: Here as well.

File: testcases/open_posix_testsuite/conformance/interfaces/free/1-1.c [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+	if (ptr2 == NULL) {
+		printf(TNAME " Test Failed, malloc() failed to allocate memory even after"
+				" freeing previously allocated memory, errno: %d\n.", errno);
+		exit(PTS_FAIL);
+	} else {
+		printf(TNAME " Test Passed, malloc() allocated memory after freeing"
+				" previously allocated memory.\n");
+		free(ptr2);
+		exit(PTS_PASS);
+	}
```
Comment (metan-ucw) [inline testcases/open_posix_testsuite/conformance/interfaces/free/1-1.c]: Useless else branch, you are calling exit at the end of the if branch.

File: testcases/open_posix_testsuite/conformance/interfaces/free/1-1.c [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+		printf(TNAME " Test Failed, malloc() is able to allocate memory before free(),"
+				" errno: %d.\n", errno);
+		exit(PTS_FAIL);
+	}
+	free(ptr1);
+	errno = 0;
+	ptr2 = malloc(1024 * 1024 * 1);
+	if (ptr2 == NULL) {
+		printf(TNAME " Test Failed, malloc() failed to allocate memory even after"
+				" freeing previously allocated memory, errno: %d\n.", errno);
```
Comment (metan-ucw) [inline testcases/open_posix_testsuite/conformance/interfaces/free/1-1.c]: Message too long. Could be "Test failed: malloc() failed allocate memory after free(), errno: %d"

File: testcases/open_posix_testsuite/conformance/interfaces/free/1-1.c [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+		exit(PTS_FAIL);
+	} else {
+		printf(TNAME " Test Passed, malloc() allocated memory after freeing"
+				" previously allocated memory.\n");
```
Comment (metan-ucw) [inline testcases/open_posix_testsuite/conformance/interfaces/free/1-1.c]: Here as well message too long.

File: testcases/open_posix_testsuite/conformance/interfaces/free/2-1.c [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+ * 		Initialise a pointer to NULL
+ *		Use free on NULL pointer
+ *		exit from child
+ *	In parent process
+ *		wait till child terminates
+ *		If child is terminated normally
+ *		test is passed.
+*/
+
+#define _XOPEN_SOURCE 700
```
Comment (metan-ucw) [inline testcases/open_posix_testsuite/conformance/interfaces/free/2-1.c]: Here as well, must not be defined for new tests.

File: testcases/open_posix_testsuite/conformance/interfaces/free/2-1.c [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+		int *ptr = NULL;
+		free(ptr);
+		exit(0);
+	} else {
+		waitpid(0, &status, pid);
+		if (WIFEXITED(status)) {
+			printf(TNAME "Test Passed\n");
+			exit(PTS_PASS);
+		}
+	}
```
Comment (metan-ucw) [inline testcases/open_posix_testsuite/conformance/interfaces/free/2-1.c]: Here as well useless else branch.

File: testcases/open_posix_testsuite/conformance/interfaces/free/2-1.c [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+	int status = 0;
+	int pid = fork();
+
+	if (pid == 0) {
+		int *ptr = NULL;
+		free(ptr);
+		exit(0);
+	} else {
+		waitpid(0, &status, pid);
+		if (WIFEXITED(status)) {
```
Comment (metan-ucw) [inline testcases/open_posix_testsuite/conformance/interfaces/free/2-1.c]: You should check exit number as well.

File: testcases/open_posix_testsuite/conformance/interfaces/free/2-1.c [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+#include <stdlib.h>
+#include <unistd.h>
+#include "posixtest.h"
+
+#define TNAME "free/2-1.c "
+
+int main(void)
+{
+
+	int status = 0;
```
Comment (metan-ucw) [inline testcases/open_posix_testsuite/conformance/interfaces/free/2-1.c]: Here as well, useless initialization.

File: testcases/open_posix_testsuite/conformance/interfaces/free/2-1.c [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+		exit(0);
+	} else {
+		waitpid(0, &status, pid);
+		if (WIFEXITED(status)) {
+			printf(TNAME "Test Passed\n");
+			exit(PTS_PASS);
+		}
+	}
+	printf(TNAME "Test Failed\n");
+	exit(PTS_FAIL);
```
Comment (metan-ucw) [inline testcases/open_posix_testsuite/conformance/interfaces/free/2-1.c]: This is pretty bad failure message, you should at least print the status integer, ideally you should decipher if the test has ended with non-zero exit status, was killed by signal, etc.

File: testcases/open_posix_testsuite/conformance/interfaces/free/1-1.c [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+
+/*
+ * assertion:
+ *
+ *	The function will cause the space pointed to by ptr to be deallocated.
+ *
+ * method:
+ *	-Set rlimit equals to 1024 * 1024 * 2.
+ *	-Allocate memory to pointer_1 of size 1024 * 1024 *1.
+ *	-Alloacte memory for pointer_2 of size 1024 *1024 *1.
```
Comment (MofX) [inline testcases/open_posix_testsuite/conformance/interfaces/free/1-1.c]: Typo: alloacte

**Pull request closed without merging**